node_tree.py

```diff
@@ -131,17 +137,38 @@ def wrapper_update(self, context):
 
     return wrapper_update
 
+# Declaring namedtuple()
+Sv_Link = collections.namedtuple('sv_link',['from_node_id', 'to_node_id', 'from_socket_id', 'to_socket_id'])
 
+def get_output_socket_id(socket):
+    if socket.node.bl_idname == 'NodeReroute':
+        if socket.node.inputs[0].is_linked:
+            return get_output_socket_id(socket.node.inputs[0].links[0].from_socket)
+        else:
+            return None, None
+    else:
+        return socket.socket_id, socket.node.node_id
 
 class SvNodeTreeCommon(object):
     '''
     Common methods shared between Sverchok node trees
     '''
-
+    sv_process: BoolProperty(name="Process", default=True, description='Process layout')
     has_changed: BoolProperty(default=False)
     limited_init: BoolProperty(default=False)
     skip_tree_update: BoolProperty(default=False)
-
+    configuring_new_node: BoolProperty(name="indicate node initialization", default=False)
+    sv_node_dict = {}
+    sv_links = {}
+    sv_linked_output_sockets = {}
+    sv_linked_input_sockets = {}
+    sv_linked_inputted_nodes = {}
```

Comment:
I would suggest consider of using static class for managing Sverchok links. It will let incapsulate all related code and let read it easier.
```Python
sv_links = SvLinks
```
For example it is possible encapsulate work of getting links for current node tree:
```Python
class SvLinks:
    sv_links = dict()

    @classmethod
    def get(cls, node_tree):
        return cls.sv_links[node_tree]

node_tree.sv_links.get(node_tree)
```
`bl_links_to_sv_links`  method can be moved to this class. It can has methods - `links has changed`, `get affected nodes`.

But updates callings should remain in `SvNodeTreeCommon` class and should be explicit. I don't like how updates are made now. It should be like this:
```Python
class SvNodeTreeCommon:
    def update():
        ...
        if links_has_changed:
            process_from_nodes(get_nodes())
        ...
```

Reply:
@Durman I separate the code in different files (links.py and node_id_dict.py) to join all the related code and cache dictionaries. is truly cleaner.
From here I'm open to learn the hows and whys of encapsulating it in classes with inheritances and classmethods and so. (To be honest I am pretty new this topics so i have just used them when i didn't find another way)

Reply:
Key to understanding using of classes is that the main their purpose is to combine data with methods. Best method is method without arguments. I would suggest to find good book about OOP.

Reply:
Modules are also a good place to combine data and methods. There are many node-trees and many nodes but just one place to 'rule' them all so i don't seed the point in making it a class that has to be initiated, it can be perfectly one module that starts when sverchok is loaded

Reply:
Modules looks like can be used as static classes but their are not the same. There are some downsides of using modules in this way:
- You have to create global variables. It is common knowledge that it is better to avoid using them. 
- You have to import each method separately (at least you do this). 
- It is harder to split module into two. 
- Module can't inherit functionality of other modules.
- It is simpler to redo static class to normal class for using its instances rather then redo module in this way.
- And probably most important thing you can't assign functionality of module to other classes fields. It would be more comfortable to have access to this functionality via `node_tree.sv_links.some_method()`

Reply:
> You have to create global variables. It is common knowledge that it is better to avoid using them.

the sv_links is a global dictionary as the socket_data_cache affects all sverchok so it makes sense to be a global variable.

> You have to import each method separately (at least you do this).

Everything stays with in the sverchok.core and sverchok.nodetree so is not a big deal to import the at the beginning 

> Module can't inherit functionality of other modules.

with a simple import at the begining you can have whatever functionality you wish. I dont think we are going to need a sv_links_common and many sv_links_classes that need to inherit this 

> It is simpler to redo static class to normal class for using its instances rather then redo module in this way.

There should be no instances of this code, is unique for all nodetrees

> And probably most important thing you can't assign functionality of module to other classes fields. It would be more comfortable to have access to this functionality via node_tree.sv_links.some_method() but instead you have to import this module everywhere where you would like to use it.

you can use sverchok.core.links.some_method(node_tree,...) in the not-so-many places you may need to use it

Reply:
> the sv_links is a global dictionary as the socket_data_cache affects all sverchok so it makes sense to be a global variable.

`sv_links` does not belong to all Sverchok. They belongs to node trees and nobody else. Socket data catch belongs to sockets. It is quite obvious. Making global in this way is mistake. We already has `update system` based on global variables. Nobody dared to improve it all years.

I have just remembered the words from legend movie:
"Unfortunately no one can be told what the Marix is. You have to see it to yourself."
So you just have to try using classes by yourself.

Reply:
@Durman excuse me for resisting your suggestions. I understand we are both here for a greater good (and also to have fun) We both give our best and is good to hear voice of someone trying to help (specially in a topic they know). The ego resists, but later understands.  So thanks again for the feedback and for pushing Sverchok to be better.

This said I don't mind to keep on improving this PR to see if it can achieve all our needs.

I have refactored the code following your suggestions and it would be  very nice if we could keep on the conversation about it